File: client/android/app/src/main/java/com/solutioncrafts/safeme/ui/features/viewmodels/TripGuardViewModel.kt
package com.solutioncrafts.safeme.ui.features.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solutioncrafts.safeme.data.network.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.util.*

class TripGuardViewModel : ViewModel() {
    private val _tripId = MutableStateFlow<String?>(null)
    val tripId: StateFlow<String?> = _tripId

    private val _currentRisk = MutableStateFlow(0.0)
    val currentRisk: StateFlow<Double> = _currentRisk

    private val _isMonitoring = MutableStateFlow(false)
    val isMonitoring: StateFlow<Boolean> = _isMonitoring

    private val _anomalyDetected = MutableStateFlow(false)
    val anomalyDetected: StateFlow<Boolean> = _anomalyDetected

    private var telemetryJob: Job? = null

    fun startMonitoring(city: String, transportType: String) {
        viewModelScope.launch {
            try {
                val startRequest = StartTripRequest(city, transportType)
                val response = NetworkClient.apiService.startTrip(startRequest)
                _tripId.value = response.tripId
                _currentRisk.value = response.initialRiskScore
                _isMonitoring.value = true
                startTelemetryLoop()
            } catch (e: Exception) {
                // handle error
            }
        }
    }

    private fun startTelemetryLoop() {
        telemetryJob?.cancel()
        telemetryJob = viewModelScope.launch {
            while (_isMonitoring.value) {
                delay(10000) // Send telemetry every 10 seconds
                sendTelemetry()
            }
        }
    }

    private suspend fun sendTelemetry() {
        val tripId = _tripId.value ?: return
        try {
            val report = TelemetryReport(
                lat = 9.0765 + (Math.random() - 0.5) * 0.01,
                lng = 7.3986 + (Math.random() - 0.5) * 0.01,
                timestamp = Date().toString()
            )
            val batch = TelemetryBatchRequest(
                tripId = tripId,
                user_id_hash = "simulated_hash",
                reports = listOf(report)
            )
            val response = NetworkClient.apiService.sendTelemetry(batch)
            _currentRisk.value = response.riskScore
            if (response.anomalyDetected) {
                _anomalyDetected.value = true
            }
        } catch (e: Exception) {
            // handle error
        }
    }

    fun triggerPanic() {
        viewModelScope.launch {
            val tripId = _tripId.value ?: return@launch
            try {
                NetworkClient.apiService.triggerPanic(
                    PanicRequest(tripId, 9.0765, 7.3986)
                )
                // Panic triggered
            } catch (e: Exception) {
                // handle error
            }
        }
    }

    fun stopMonitoring() {
        _isMonitoring.value = false
        telemetryJob?.cancel()
        _tripId.value = null
    }
}

File: client/ios/safeme/Features/ViewModels/TripGuardViewModel.swift
import Foundation
import Combine

struct StartTripRequest: Encodable {
    let city: String
    let transportType: String
}

struct TripResponse: Decodable {
    let tripId: String
    let initialRiskScore: Double
}

struct TelemetryReport: Encodable {
    let lat: Double
    let lng: Double
    let timestamp: String
}

struct TelemetryBatchRequest: Encodable {
    let tripId: String
    let user_id_hash: String
    let reports: [TelemetryReport]
}

struct TelemetryResponse: Decodable {
    let anomalyDetected: Bool
    let riskScore: Double
}

struct PanicRequest: Encodable {
    let tripId: String
    let lat: Double
    let lng: Double
}

@MainActor
class TripGuardViewModel: ObservableObject {
    @Published var tripId: String?
    @Published var currentRisk: Double = 0.0
    @Published var isMonitoring = false
    @Published var anomalyDetected = false
    
    private var timer: Timer?
    
    func startMonitoring(city: String, transportType: String) async {
        do {
            let request = StartTripRequest(city: city, transportType: transportType)
            let endpoint = APIEndpoint(path: "/trip-guard/start", method: "POST", body: try? JSONEncoder().encode(request))
            
            let response = try await APIClient.shared.request(endpoint, responseType: TripResponse.self)
            self.tripId = response.tripId
            self.currentRisk = response.initialRiskScore
            self.isMonitoring = true
            
            startTelemetryLoop()
        } catch {
            print("Error starting trip: \(error)")
        }
    }
    
    private func startTelemetryLoop() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 10, repeats: true) { _ in
            Task {
                await self.sendTelemetry()
            }
        }
    }
    
    private func sendTelemetry() async {
        guard let tripId = self.tripId else { return }
        
        do {
            let report = TelemetryReport(
                lat: 9.0765 + (Double.random(in: -0.005...0.005)),
                lng: 7.3986 + (Double.random(in: -0.005...0.005)),
                timestamp: ISO8601DateFormatter().string(from: Date())
            )
            let batch = TelemetryBatchRequest(
                tripId: tripId,
                user_id_hash: "simulated_hash_ios",
                reports: [report]
            )
            
            let endpoint = APIEndpoint(path: "/trip-guard/telemetry", method: "POST", body: try? JSONEncoder().encode(batch))
            let response = try await APIClient.shared.request(endpoint, responseType: TelemetryResponse.self)
            
            self.currentRisk = response.riskScore
            if response.anomalyDetected {
                self.anomalyDetected = true
            }
        } catch {
            print("Error sending telemetry: \(error)")
        }
    }
    
    func triggerPanic() async {
        guard let tripId = self.tripId else { return }
        
        do {
            let request = PanicRequest(tripId: tripId, lat: 9.0765, lng: 7.3986)
            let endpoint = APIEndpoint(path: "/trip-guard/panic", method: "POST", body: try? JSONEncoder().encode(request))
            _ = try await APIClient.shared.request(endpoint, responseType: [String: String].self)
        } catch {
            print("Error triggering panic: \(error)")
        }
    }
    
    func stopMonitoring() {
        isMonitoring = false
        timer?.invalidate()
        tripId = nil
    }
}
